Rewrite throwing to use same rules as shooting

Coolthulhu opened the pull request:
Throwing used weird system, totally split from other ranged attacks, a bit like an inlined subset of shooting ranged attacks. For example, directly applying damage, skipping monster dodge and size and skipping monsters' abilities such as damage resistance or on-hit effects.

I rewrote throwing to use the same rules as shot projectiles. This should greatly improve maintainability, should someone want to add all the features in the comments, like dodging projectiles.

From player's perspective:
- Thrown items now fly through the air like bullets (didn't code them to display proper sprite, though - red '*' has to be enough for now)
- Spitters and corrosives now show an animated bullet too. Also, dynamics of their bullets are more player-like
- Thrown acid bombs will explode into acid even if the player doesn't activate them first
- Projectiles embed themselves into targets if they cut/stab more than they bash. Affects both thrown knives and arrows/bolts
- Nerfed shrapnel a bit: it now deals 50% bashing and 50% stabbing damage, meaning armor counts twice against it.

---

From the code side:
- Rewrote the throwing function. Moved from `game` to `player. NPCs will now use the same rules as player. It uses`Creature::projectile_attack` to deal the attack.
- Removed most of the `projectile` class, because I hate ancient placeholders with no implementation.
- Removed minor "caches" from `projectile` class: `wide` or `drops`. "WIDE" is now a flag, `drops` is determined by `unique_ptr<item>` being not nullptr
- Ranged attacks now take and return a struct to describe the attack. I'm not sure if one or two structs would be better (one before, one after the attack). Went with one, because most of the fields would be shared.
- Thrown items can now be activated on drop from projectile. Currently only acid bombs do it, but coding thrown impact grenades would be easy now.
- Rewrote shrapnel code to create a fake NPC and have it "fire" the shrapnel around. Less copypasted code.

TODO:
- [x] "The debug monster is hit by/dodges Shrapnel's projectile!"
- [x] Wrap the `unique_ptr<item>` into a less transparent class
- [x] Bring back shrapnel animation skip, so that ANFO doesn't take 10 seconds to animate
- [x] Fix shrapnel crashing the game

Barhandar commented:
Does that mean throwing will no longer allow lowly survivors to defend themselves against high-dodge enemies?

Coolthulhu commented:
It should still work, just not be an instant-hit (old rules had same thrown accuracy vs. hulks and dogs).

High-dodge monsters now no longer require getting naked early on, though, so melee should be more viable against them.

BevapDin commented on data/json/items/tools.json:
```diff
@@ -3223,6 +3223,7 @@
     "description": "This is a glass vial, split into two chambers.  The divider is removable, which will cause the chemicals to mix.  If this mixture is exposed to air (perhaps by throwing the vial) they will spill out as a pool of potent acid.",
     "price": 5000,
     "material": "glass",
+    "flags": "PROCESS_ON_DROP",
```
"ON_DROP" indicates (to me) it will be processed when the character drops it (as with the drop items command). Sounds more like this should be "ON_THROW".

Coolthulhu replied:
I wanted it to be something more generic, so that it could be used both for ammo and thrown items. Also, "ON_THROW" could suggest that it is processed in mid-air.
"ON_HIT" could work.

illi-kun replied:
Please don't forget to add the description of the introduced flag(s) to the documentation.

BevapDin commented on src/damage.h:
```diff
@@ -90,28 +85,22 @@ struct resistances {
+
+    projectile();
+    projectile( const projectile& );
+    projectile& operator=( const projectile& );
```
Do we need a copy constructor and assignment operator? Maybe the move constructor and move assignment are enough, if so, they can be compiler generated (works fine for `unique_ptr`).

Coolthulhu replied:
Assignment happens in the code, though. For example, bouncing projectiles and shrapnel.

BevapDin commented on src/ranged.cpp:
```diff
@@ -593,36 +602,67 @@ void player::fire_gun( const tripoint &targ_arg, bool burst )
+    // We'll be constructing a projectile
+    // Start by copying the item into it
+    projectile proj;
+    proj.drop = std::unique_ptr<item>( new item( to_throw ) );
```
Usually, I use `drop.reset( new item( ... ) )`, but I also that it makes no difference beside being shorter (-:

BevapDin commented on src/ranged.cpp:
```diff
@@ -630,205 +670,109 @@ void game::throw_item( player &p, const tripoint &target, item &thrown,
+        "iron", "steel"
+    }};
+    */
+    std::vector<std::string> ferric;
```
If it has a fixed size and is not changed during runtime, why not make it a `std::array`?

Coolthulhu replied:
Can array be passed as vector&?

BevapDin replied:
I've just seen, you use it as argument to `item::made_of_any`, which requires a `std::vector`, so it must be a vector. Strange, the static vector should work fine, it's already used that way in `iuse::solder_weld`.

BevapDin commented on src/ranged.cpp:
```diff
@@ -1531,3 +1476,66 @@ void splatter( const std::vector<tripoint> &trajectory, int dam, const Creature
+            }
+        }
+    } else {
+        mon->add_item( dropped_item );
```
This requires a message. It's very confusing for experienced players that expect the item to appear under the monster.

The PR works technically fine (no idea about the numbers and balancing).

Coolthulhu commented:
OK, this should be ready.

Changed the flag to "ACT_ON_RANGED_HIT". Should be unambiguous.

Added descriptions of the flags to documentation.

Refactored the projectile item drop: the pointer is not exposed, instead it is accessed by a getter/setter. For a case where it is null, a static item is returned, for which `is_null()` is true.

Also fixed minor bugs:
* Unintended targets on the line of fire were hit more accurately if the player had really bad accuracy vs. the intended target. Shooting far behind intended target could boost accuracy.
* High strength would lower accuracy with thrown weapons, high weight of the item would increase accuracy.

BevapDin commented on src/monattack.cpp:
```diff
@@ -1299,77 +1247,21 @@ void mattack::spit_sap(monster *z, int index)
-    target->check_dead_state();
+    projectile proj;
+    proj.speed = 10;
+    proj.proj_effects.insert( "APPLY_SAP" );
```
The flag, it does nothing. There is no other reference to it anywhere in the code or in JSON data.

Coolthulhu replied:
Oops, forgot to implement it. Pushing an implementation now.
The balancing changed, but that's a rare ability so a small buff/nerf here shouldn't be a problem.

BevapDin commented:
There's a consistent problem here: when `drop_or_embed_projectile` is called, the monster is sometimes already dead, which means its `die` method has already been run, which means its inventory is already dropped. Adding the thrown item to the inventory means it's not dropped (`die` is run exactly once).

A solution would be to handle a dead target the same as no target monster in `drop_or_embed_projectile` and simply drop the item. This also gets rid of the message sequence "The foo dies!" followed by "The bar embeds in the foo!"

BevapDin commented:
Not really related, but `mattack::spit_sap` calls `monster::can_act`, which returns false if the monster has the IMMOBILE flag. biollant monsters have that flag and as a result, their special attack (`spit_sap`) is effectively ignored. The same problem might exist with other monster attack that use `can_act`.
